python/cudf/cudf/core/buffer.py

```diff
@@ -2,111 +2,226 @@
-        object is kept in this Buffer.
+        Python object to which the lifetime of the memory allocation is tied.
+        A reference to this object is kept in the returned Buffer.
     """
 
     _ptr: int
     _size: int
     _owner: object
 
     def __init__(
-        self, data: Any = None, size: int = None, owner: object = None
+        self, data: Union[int, Any], *, size: int = None, owner: object = None
     ):
-        if isinstance(data, Buffer):
-            self._ptr = data._ptr
-            self._size = data.size
-            self._owner = owner or data._owner
-        elif isinstance(data, rmm.DeviceBuffer):
-            self._ptr = data.ptr
-            self._size = data.size
-            self._owner = data
-        elif hasattr(data, "__array_interface__") or hasattr(
-            data, "__cuda_array_interface__"
-        ):
-            self._init_from_array_like(data, owner)
-        elif isinstance(data, memoryview):
-            self._init_from_array_like(np.asarray(data), owner)
-        elif isinstance(data, int):
-            if not isinstance(size, int):
-                raise TypeError("size must be integer")
+        if isinstance(data, int):
+            if size is None:
+                raise ValueError(
+                    "size must be specified when `data` is an integer"
+                )
+            if size < 0:
+                raise ValueError("size cannot be negative")
             self._ptr = data
             self._size = size
             self._owner = owner
-        elif data is None:
-            self._ptr = 0
-            self._size = 0
-            self._owner = None
         else:
-            try:
-                data = memoryview(data)
-            except TypeError:
-                raise TypeError("data must be Buffer, array-like or integer")
-            self._init_from_array_like(np.asarray(data), owner)
+            if size is not None or owner is not None:
+                raise ValueError(
+                    "`size` and `owner` must be None when "
+                    "`data` is a buffer-like object"
+                )
+
+            # `data` is a buffer-like object
+            buf: Any = data
+            if isinstance(buf, rmm.DeviceBuffer):
+                self._ptr = buf.ptr
+                self._size = buf.size
+                self._owner = buf
+                return
+            iface = getattr(buf, "__cuda_array_interface__", None)
+            if iface:
+                ptr, size = get_ptr_and_size(iface)
+                self._ptr = ptr
+                self._size = size
+                self._owner = buf
+                return
+            ptr, size = get_ptr_and_size(np.asarray(buf).__array_interface__)
+            buf = rmm.DeviceBuffer(ptr=ptr, size=size)
+            self._ptr = buf.ptr
+            self._size = buf.size
+            self._owner = buf
+
+    def __getitem__(self, key: Union[int, slice]) -> Buffer:
+        if isinstance(key, int):
```

Comment:
What is the use case of being able to call `buffer[0]`? Do we ever want to fetch buffers of a single byte? If not, should we disable indexing by `int` and only allow slices?

This question came to me because when `__getitem__` is implemented for collection-like types, it's unusual to return the same type of container when indexing a specific value. Normally the value itself (rather than another, smaller container) is returned. However, a device-to-host copy of a single byte would also be a little weird.

Reply:
Good point, I have removed support of `int`.